### PR TITLE
Warn on broken steps, use yt-dlp to avoid youtube-dl errors, and don't crash on bad UTF-8

### PR DESCRIPTION
**Pull request by turian**
# Summary

* If some step is broken, we issue a warning instead of crashing.
* We use yt-dlp instead of youtube-dl by default.
* On bad unicode, we don't crash.

# Quickest workaround for many people, until this is merged

Add this to ArchiveBox.conf:
```
YOUTUBEDL_BINARY=/usr/bin/yt-dlp
```

If that doesn't work, you can use my Docker `turian/archivebox:kludge-984-UTF8-bug` which includes this patch, instead of `archivebox/archivebox` for now.

@jgoerzen I think you also said this bug was a showstopper for you

# Try it out

This finally works with this patch:

`archivebox add 'https://www.ashra.com/news.php?m=A'`

# Related issues

Should close these issues:
* UnicodeDecodeError: https://github.com/ArchiveBox/ArchiveBox/issues/1014 and https://github.com/ArchiveBox/ArchiveBox/issues/984 https://github.com/ArchiveBox/ArchiveBox/issues/999
* Not able to save YouTube videos which might be UnicodeDecodeError but are probably also `youtube-dl` errors: https://github.com/ArchiveBox/ArchiveBox/issues/991 https://github.com/ArchiveBox/ArchiveBox/issues/998 

# Changes these areas

- [X] Bugfixes
- [ ] Feature behavior
- [ ] Command line interface
- [ ] Configuration options
- [ ] Internal architecture
- [ ] Snapshot data layout on disk

# `youtube-dl` -> `yt-dlp`

So I began by changing the `archive_link` from a crashing exception to a warning. This gave me better diagnostics than the exceptions you see in the issues above. I observed the following:

```
       Extractor failed:
             Failed to save media
        Run to see full output:
            cd /home/archivebox/archivebox-turian/ArchiveBox/pipdata/archive/1663011964.432896;
            youtube-dl --write-description --write-info-json --write-annotations --write-thumbnail --no-call-home --write-sub --all-subs --write-auto-sub --convert-subs=srt --yes-playlist --continue --ignore-errors --no-abort-on-error --geo-bypass --add-metadata --max-filesize=750m https://www.ashra.com/news.php?m=A
```

Inspecting that, I saw:

```
$ youtube-dl --write-description --write-info-json --write-annotations --write-thumbnail --no-call-home --write-sub --all-subs --write-auto-sub --convert-subs=srt --yes-playlist --continue --ignore-errors --no-abort-on-error --geo-bypass --add-metadata --max-filesize=750m https://www.ashra.com/news.php?m=A
Usage: youtube-dl [OPTIONS] URL [URL...]

youtube-dl: error: no such option: --no-abort-on-error
```

Oups. `youtube-dl` doesn't have `--no-abort-on-error`. I think it only has `--abort-on-error` but I'm not entirely sure what the default behavior is.

I switched [youtube-dl](https://github.com/ytdl-org/youtube-dl) to [yt-dlp](https://github.com/yt-dlp/yt-dlp) as the default, which is a more actively maintained fork that pulls upstream from youtube-dl constantly. It also has option `--no-abort-on-error` so now the media download works.

It operates essentionally identically (and FASTER) to youtube-dl.

The above youtube-dl options are present, with the following caveats:
* We aren't doing this, but "When --embed-subs and --write-subs are used together, the subtitles are written to disk and also embedded in the media file. You can use just --embed-subs to embed the subs and automatically delete the separate file."
* We aren't doing this, but "--add-metadata attaches the infojson to mkv files in addition to writing the metadata when used with --write-info-json. Use --no-embed-info-json or --compat-options no-attach-info-json to revert this"
* `--write-annotations`: "No supported site has annotations now"

I haven't pushed yt-dlp default changes to the submodules yet:
`deb_dist, brew_dist, docker, docs, etc/ArchiveBox.conf.default, pip_dist`
but I'd like to. I would even go so far as to deprecate `youtube-dl`, but perhaps that's too radical for some people.

# UnicodeDecodeError fixes

A common complaint. In `media.py`, we used to have:
```
text_file.read_text(encoding='utf-8').strip()
```
This is strict and errors cause a crash. I have changed the behavior to [xmlcharrefreplace](https://github.com/ArchiveBox/ArchiveBox/pull/1026/files#diff-9025388b81475556336588c7ee9ba7dbee7e2afb73c962c1fce3d81bc6ca9dd8R77) ("Characters not supported by the encoding are replaced with the appropriate XML character reference &#nnn;.") which seemed the best to me, but you can read about other options in my comment there or in more detail at Python's [open](https://docs.python.org/3/library/functions.html#open) function documentation.

My preferred workaround, because sometimes things say they are utf-8 but are actually a different encoding, would be this:
1) Try utf-8 in strict mode. If clean, then good.
2) If error, use [chardet](https://chardet.readthedocs.io/en/latest/usage.html#basic-usage) to guess the encoding. (It's a little slow, so it's better as a fallback option.) Attempt to decode to Unicode using the guessed encoding in strict mode. If clean, then good.
3) Worst case: Either use utf-8 or the highest confidence `chardet` guessed encoding, and decode it using `xmlcharrefreplace` for errors.

This would be a separate PR, if you approve of me including `chardet` as a pip dependency here and in all submodules (same as with yt-dlp).

# Postscript

I ran `flake8` and caught all the flakes in the code I introduced, as far as I know. Since I am a new committer, CI/CD won't run for me here.

### CODE REVIEW

File: archivebox/extractors/__init__.py [hunk 1]
```diff
@@ -126,11 +128,28 @@ def archive_link(link: Link, overwrite: bool=False, methods: Optional[Iterable[s
...
+                with open(ERROR_LOG, "a", encoding='utf-8') as f:
+                    command = ' '.join(sys.argv)
+                    ts = datetime.now(timezone.utc).strftime('%Y-%m-%d__%H:%M:%S')
+                    f.write(("\n" + 'Exception in archive_methods.save_{}(Link(url={})) command={}; ts={}'.format(
+                        method_name,
+                        link.url,
+                        command,
+                        ts
+                    ) + "\n"))
+                    #f.write(f"\n> {command}; ts={ts} version={config['VERSION']} docker={config['IN_DOCKER']} is_tty={config['IS_TTY']}\n")
```
Comment (turian) [inline archivebox/extractors/__init__.py:152]: Should there be a `break` here to exit the loop?

**Comment by notevenaperson** [comment]
+1 for yt-dlp

File: setup.py [hunk 2]
```diff
@@ -42,6 +42,7 @@
     "django-extensions>=3.0.3",
     "dateparser>=1.0.0",
     "youtube-dl>=2021.04.17",
+    "yt-dlp>=2021.4.11",
```
Comment (turian) [inline setup.py:45]: I arbitrarily picked this version number to be the first datestamp ahead of the youtube-dl version.

**Comment by turian** [comment]
A few more notes:
* `yt-dlp` is so much faster than `youtube-dl` that you get throttled if you download all the autogenerated subtitles. So I simply disabled them, and only generate "real" subtitles and captions.

File: archivebox/config.py [hunk 3]
```diff
@@ -144,12 +144,19 @@
...
                                                                 '--convert-subs=srt',
                                                                 '--yes-playlist',
                                                                 '--continue',
-                                                                '--ignore-errors',
+                                                                # This flag doesn't exist in youtube-dl
+                                                                # only in yt-dlp
                                                                 '--no-abort-on-error',
+                                                                # --ignore-errors must come AFTER
+                                                                # --no-abort-on-error
+                                                                # https://github.com/yt-dlp/yt-dlp/issues/4914
```
Comment (turian) [inline archivebox/config.py:158]: The order of these flags is---confusingly---quite important.

https://github.com/yt-dlp/yt-dlp/issues/4914

**Comment by WakeReality** [comment]
I'm using docker-compose to run ArchiveBox. How do I get yt-dlp inside the docker container?

**Comment by pirate** [comment]
Yt-dlp is already the default in the new beta 0.4.3 `latest` builds.

File: setup.py (hunk 2 above)
Comment (pirate) [reply setup.py:45]: you can bump this to the latest version whenever you touch these files (as long as it works), no need to match older versions of chrome/youtubledl/ffmpeg/git/wget, only django and a couple other edge cases are the ones that cant be bumped

**Review by pirate — APPROVED** [review]

**Comment by turian** [comment]
boom!

**Pull request merged**